README.md

```diff
@@ -100,51 +149,140 @@ shell>  cat info.log
 
-julia> filtered_logger = FilteredLogger(yodawg_filter, global_logger());
+julia> filtered_logger = ActiveFilteredLogger(yodawg_filter, global_logger());
 
 julia> with_logger(filtered_logger) do
-		@info "Boring message"
-		@warn "Yo Dawg! it is bad"
-		@info "Another boring message"
-		@info "Yo Dawg! it is all good"
-	end
+	@info "Boring message"
+	@warn "Yo Dawg! it is bad"
+	@info "Another boring message"
+	@info "Yo Dawg! it is all good"
+end
 ┌ Warning: Yo Dawg! it is bad
 └ @ Main REPL[28]:3
 [ Info: Yo Dawg! it is all good
 ```
 
+## `EarlyFilteredLogger`
 
+The `EarlyFilteredLogger` is similar to the `ActiveFilteredLogger`,
+but it runs earlier in the logging pipeline.
+In particular it runs before the message is computed.
+It is (theoretically) useful to filter things early if creating the log message is expective.
+E.g. if it includes summary statistics of the error.
+The filter function for early filter logging only has access to the 
+`level`, `_module`, `id` and `group` fields of the log message.
+The most notable use of it is to filter based on modules,
+see the HTTP example below.
 
-# Examples
+Another example is using them to stop messages every being repeated within a given time period.
+
+```
+using Dates, Logging, LoggingExtras
+
+julia> function make_throttled_logger(period)
+         history = Dict{Symbol, DateTime}()
+         # We are going to use a closure
+         EarlyFilteredLogger(global_logger()) do log
+           if !haskey(history, log.id) || (period < now() - history[log.id])
+             # then we will log it, and update record of when we did
+             history[log.id] = now()
+             return true
+           else
+             return false
+           end
+         end
+       end
+make_throttled_logger (generic function with 1 method)
+
+julia> throttled_logger = make_throttled_logger(Second(3));
+
+julia> with_logger(throttled_logger) do
+         for ii in 1:10
+           sleep(1)
+           @info "It happen" ii
+         end
+       end
+┌ Info: It happen
+└   ii = 1
+┌ Info: It happen
+└   ii = 4
+┌ Info: It happen
+└   ii = 7
+┌ Info: It happen
+└   ii = 10
+```
+
+## `MinLevelLogger`
+This is basically a special case of the early filtered logger,
+that just checks if the level of the message is above the level specified when it was created.
+
+## `TransformerLogger`
+The transformer logger allows for the modification of log messages.
+This modification includes such things as its log level, and content,
+and all the other arguments passed to `handle_message`.
+
+When constructing a `TransformerLogger` you pass in a tranformation function,
+and a logger to be wrapped.
+The  transformation function takes a named tuple containing all the log message fields,
+and should return a new modified named tuple.
+
+A simple example of its use is truncating messages.
+
+```
+julia> using Logging, LoggingExtras
+
+julia> truncating_logger  = TransformerLogger(global_logger()) do log
```

Comment:
We need to do this bundling of all the keys and values together, I like it.

One question - what happens if someone does
```
@info "asdf" message=123
```

I was wondering whether we should have a `LogRecord` type for this instead of a bare named tuple? The annoying thing is that it makes field access less convenient for the arbitrary key value pairs.

Reply:
Oh never mind, I see that you end up with `log.message` and `log.kwargs.message` for this.  That makes sense and is the same as you'd get if we were to have a `LogRecord` type.

Reply:
Reason for a `NamedTuple` rather than a `struct` was so I could splat it back into the full form.
If we added a `LogRecord` struct to `Core.Logging` (and for backwards compat made the single args get wrapped into it),
Then I think that would be fine.

Reply:
Yep I think we need to do something like that.